src/elfpy/user.py

```diff
@@ -4,73 +4,134 @@
+    token_in_wallet: dict = field(default_factory=dict)
+    base_in_protocol: dict = field(default_factory=dict)
+    token_in_protocol: dict = field(default_factory=dict)
+    effective_price: float = field(init=False)  # calculated after init, only for transactions
+
+    def __post_init__(self):
+        """post initialization function"""
+        # check if this represents a trade (one side will be negative)
+        total_tokens = sum(list(self.token_in_wallet.values()))
+        if self.base_in_wallet < 0 or total_tokens < 0:
+            self.effective_price = total_tokens / self.base_in_wallet
+
+    def __getitem__(self, key):
+        getattr(self, key)
+
+    def __setitem__(self, key, value):
+        setattr(self, key, value)
+
+    def __str__(self):
+        output_string = ""
+        for key, value in vars(self).items():
+            if value:  #  check if object exists
+                if value != 0:
+                    output_string += f" {key}: "
+                    if isinstance(value, float):
+                        output_string += f"{float_to_string(value)}"
+                    elif isinstance(value, list):
+                        output_string += "[" + ", ".join([float_to_string(x) for x in value]) + "]"
+                    elif isinstance(value, dict):
+                        output_string += "{" + ", ".join([f"{k}: {float_to_string(v)}" for k, v in value.items()]) + "}"
+                    else:
+                        output_string += f"{value}"
+        return output_string
+
+
 class User:
     """
     Implements abstract classes that control user behavior
     user has a budget that is a dict, keyed with a date
     value is an inte with how many tokens they have for that date
     """
 
-    def __init__(self, market, rng, budget=0, verbose=False):
+    def __init__(self, market, rng, wallet_address, verbose, **kwargs):
         """
         Set up initial conditions
         """
         self.market = market
-        self.budget = budget
-        assert self.budget >= 0, f"ERROR: budget should be initialized (>=0), but is {self.budget}"
-        self.wallet = self.UserWallet(base_in_wallet=self.budget)
         self.rng = rng
+        self.wallet_address = wallet_address
         self.verbose = verbose
         self.last_update_spend = 0
+        self.product_of_time_and_base = 0
         self.weighted_average_spend = 0
-
-    @dataclass
-    class UserWallet:
-        """user wallet store"""
-
-        # pylint: disable=missing-function-docstring
-        base_in_wallet: float = 0
-        token_in_wallet: dict = field(default_factory=dict)
-        base_in_protocol: dict = field(default_factory=dict)
-        token_in_protocol: dict = field(default_factory=dict)
-
-        def __getitem__(self, key):
-            return getattr(self, key)
-
-        def __setitem__(self, key, value):
+        self.position_list = []
+        for key, value in kwargs.items():
```

Comment:
I added this and the `**kwargs` in the signature, but it is a really bad coding pattern (creates easy opportunity for attacks, also v opaque in terms of what args are accepted). I don't think we need them, so can you delete?